src/modules/storage/LedgerStorage.ts

```diff
@@ -3040,6 +3413,143 @@ export class LedgerStorage extends Storages {
         return this.query(sql, [address]);
     }
 
+    /**
+     * Get all proposals
+     * @param limit Maximum record count that can be obtained from one query
+     * @param page The number on the page, this value begins with 1
+     * @returns returns the Promise with requested data
+     * and if an error occurs the .catch is called with an error.
+     */
+    public getProposals(limit: number, page: number): Promise<any> {
+        let sql = `
+        SELECT 
+            P.proposal_id,
+            P.title,
+            P.type,
+            P.funding_amount,
+            P.submit_time,
+            P.voting_start_height,
+            P.voting_end_height,
+            P.status,
+            U.proposer_name 
+        FROM proposals AS P 
+        INNER JOIN proposer AS U 
+        ON(P.proposer_id = U.proposer_id)
+        LIMIT ? OFFSET ?`;
+        return this.query(sql, [limit, limit * (page - 1)]);
+    }
+
+    /**
+     * Get proposal by id
+     * @param proposal_id Id of the proposal
+     * @returns returns the Promise with requested data
+     * and if an error occurs the .catch is called with an error.
+     */
+    public getProposalById(proposal_id: string): Promise<any> {
+        let sql = `
+        SELECT 
+            P.proposal_id,
+            P.title,
+            P.detail,
+            P.tx_hash,
+            P.fee_tx,
+            P.funding_amount,
+            P.vote_fee,
+            P.status,
+            P.type,
+            P.voting_start_height,
+            P.voting_start,
+            P.voting_end_height,
+            P.voting_end,
+            P.submit_time,
+            Q.proposer_name 
+        FROM proposals AS P
+        INNER JOIN proposer AS Q
+        ON(P.proposer_id = Q.proposer_id)
+        WHERE P.proposal_id = ?`;
+        return this.query(sql, [proposal_id]);
+    }
+
+    /**
+     * Get all votes for proposal
+     * @param proposal_id Id of the proposal
+     * @returns returns the Promise with requested data
+     * and if an error occurs the .catch is called with an error.
+     */
+    public getProposalVotesAPI(proposal_id: string, limit: number, page: number): Promise<any> {
+        let sql = `
+        SELECT 
+            voter_utxo,
+            voter_address,
+            sequence,
+            tx_hash,
+            voting_time
+        FROM vote
+        WHERE proposal_id = ?
+        LIMIT ? OFFSET ?
```

Comment:
`ballot_answer` Don't we need to add relevant results?
We don't know ballot_answer before counting, but It can inquire after counting.
If you don't know the result yet, just mark it as an empty string.

Reply:
resloved